File: common/test/http.py
# ovirt-imageio
# Copyright (C) 2015-2018 Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

"""
http helpers.
"""

from __future__ import absolute_import
from __future__ import print_function

import errno
import logging

from pprint import pprint

from six.moves import http_client

from ovirt_imageio_common import config
from ovirt_imageio_common import pki
from ovirt_imageio_common import uhttp

log = logging.getLogger("test")


def connection():
    return http_client.HTTPSConnection(
        config.images.host,
        config.images.port,
        pki.key_file(config),
        pki.cert_file(config))


def get(uri, headers=None):
    return request("GET", uri, headers=headers)


def put(uri, body, headers=None):
    return request("PUT", uri, body=body, headers=headers)


def patch(uri, body, headers=None):
    return request("PATCH", uri, body=body, headers=headers)


def options(uri):
    return request("OPTIONS", uri)


def request(method, uri, body=None, headers=None):
    con = connection()
    try:
        con.request(method, uri, body=body, headers=headers or {})
    except EnvironmentError as e:
        if not (e.errno == errno.EPIPE and body):
            raise
        log.warning("Error sending request: %s", e)
    return response(con)


def raw_request(method, uri, body=None, headers=None):
    """
    Use this to send bad requests - this will send only the headers set in
    headers, no attempt is made to create a correct request.
    """
    con = connection()
    con.putrequest(method, uri)
    if headers:
        for name, value in headers.items():
            con.putheader(name, value)
    con.endheaders()

    try:
        if body:
            con.send(body)
    except EnvironmentError as e:
        if e.errno != errno.EPIPE:
            raise
        log.warning("Error sending body: %s", e)

    return response(con)


def local(method, uri, body=None, headers=None):
    return unix_request(
        config.images.socket, method, uri, body=body, headers=headers)


def unix_request(socket, method, uri, body=None, headers=None):
    con = uhttp.UnixHTTPConnection(socket)
    try:
        con.request(method, uri, body=body, headers=headers or {})
    except EnvironmentError as e:
        if not (e.errno == errno.EPIPE and body):
            raise
        log.warning("Error sending request: %s", e)

    return response(con)


def response(con):
    res = con.getresponse()
    pprint((res.status, res.reason, res.version))
    pprint(res.getheaders())
    return res

File: go/http/http.go
// ovirt-imageio
// Copyright (C) 2021 Red Hat, Inc.
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as
// published by the Free Software Foundation; either version 2 of the
// License, or (at your option) any later version.

package http

import (
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"net/http"

	"ovirt.org/imageio"
)

// Backend exposes a disk image served by imageio server on a oVirt host.
type Backend struct {
	url     string
	client  *http.Client
	size    uint64
	extents []*imageio.Extent
}

// Connect returns a connected Backend. Caller should close the backend when
// done.
func Connect(url string) (*Backend, error) {
	tr := &http.Transport{
		// TODO: Support server certificate verification.
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},

		// Increass throughput from 400 MiB/s to 1.3 GiB/s
		// https://go-review.googlesource.com/c/go/+/76410.
		WriteBufferSize: 128 * 1024,

		// TODO: connection and read timeouts.
	}
	client := &http.Client{Transport: tr}

	// TODO: Send OPTIONS request

	return &Backend{url: url, client: client}, nil
}

// Size return image size.
func (b *Backend) Size() (uint64, error) {
	if b.size == 0 {
		// imageio does not expose the size of the image in the OPTIONS request
		// yet. The only way to get size is to get all the extents and compute
		// the size from the last extent.
		err := b.getExtents()
		if err != nil {
			return 0, err
		}
		last := b.extents[len(b.extents)-1]
		b.size = last.Start + last.Length
	}
	return b.size, nil
}

// Extents returns all image extents. Imageio server does not support getting
// partial extent yet.
func (b *Backend) Extents() (imageio.ExtentsResult, error) {
	if len(b.extents) == 0 {
		if err := b.getExtents(); err != nil {
			return nil, err
		}
	}
	return imageio.NewExtentsWrapper(b.extents), nil
}

// Close closes the connection to imageio server.
func (b *Backend) Close() {
	b.client.CloseIdleConnections()
}

func (b *Backend) getExtents() error {
	res, err := b.client.Get(b.url + "/extents")
	if err != nil {
		return err
	}
	defer res.Body.Close()

	if res.StatusCode != http.StatusOK {
		return readServerError(res)
	}

	body, err := io.ReadAll(res.Body)
	if err != nil {
		return fmt.Errorf("Cannot get extents: %s", err)
	}

	err = json.Unmarshal(body, &b.extents)
	if err != nil {
		return fmt.Errorf("Cannot get extents: %s", err)
	}

	return nil
}

func readServerError(res *http.Response) error {
	reason, err := io.ReadAll(res.Body)
	if err != nil {
		return fmt.Errorf("Error reading response: %s", err)
	}
	return fmt.Errorf("Server error: %s", reason)
}